### PR TITLE
Add PHP 8 support

### PR DESCRIPTION
**Pull request by smoench**
Waiting for:

- [x]  https://github.com/oradwell/covers-validator to be PHP 8 compatible
- [x] fix  e2e_0211 - 8.0
- [x]  fix e2e_027
- [x]  fix e2e_028 - 7.3 (@smoench)
- [x]  fix e2e_028 - 8.0 (@smoench)

### CODE REVIEW

**Comment by smoench** [comment]
@theofidry @villfa It seems `e2e_020` fails with composer v2. With composer v1 it works but composer v1 is not compatible with PHP 8.0 yet. Can you have look?

<details><pre>
composer --working-dir=fixtures/set020-infection install
 146/146 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%Installing dependencies from lock file (including require-dev)
Verifying lock file contents can be installed on current platform.
Package operations: 29 installs, 0 updates, 0 removals
  - Downloading infection/extension-installer (0.1.1)
  - Downloading thecodingmachine/safe (v1.3.3)
  - Downloading symfony/process (v5.2.0)
  - Downloading symfony/finder (v5.2.0)
  - Downloading symfony/filesystem (v5.2.0)
  - Downloading sanmai/pipeline (v5.1.0)
  - Downloading ondram/ci-detector (3.5.1)
  - Downloading infection/include-interceptor (0.2.4)
  - Downloading infection/abstract-testframework-adapter (0.3.1)
  - Downloading infection/infection (0.18.2)
  0/10 [>---------------------------]   0%
  1/10 [==>-------------------------]  10%
  2/10 [=====>----------------------]  20%
  3/10 [========>-------------------]  30%
  4/10 [===========>----------------]  40%
  6/10 [================>-----------]  60%
  7/10 [===================>--------]  70%
  8/10 [======================>-----]  80%
  9/10 [=========================>--]  90%
 10/10 [============================] 100%  - Installing composer/package-versions-deprecated (1.11.99.1): Extracting archive
  - Installing infection/extension-installer (0.1.1): Extracting archive
  - Installing symfony/polyfill-ctype (v1.20.0): Extracting archive
  - Installing webmozart/assert (1.9.1): Extracting archive
  - Installing webmozart/path-util (2.3.0): Extracting archive
  - Installing thecodingmachine/safe (v1.3.3): Extracting archive
  - Installing symfony/polyfill-php80 (v1.20.0): Extracting archive
  - Installing symfony/process (v5.2.0): Extracting archive
  - Installing symfony/finder (v5.2.0): Extracting archive
  - Installing symfony/filesystem (v5.2.0): Extracting archive
  - Installing symfony/polyfill-mbstring (v1.20.0): Extracting archive
  - Installing symfony/polyfill-intl-normalizer (v1.20.0): Extracting archive
  - Installing symfony/polyfill-intl-grapheme (v1.20.0): Extracting archive
  - Installing symfony/string (v5.2.0): Extracting archive
  - Installing psr/container (1.0.0): Extracting archive
  - Installing symfony/service-contracts (v2.2.0): Extracting archive
 1213/1270 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓░░]  95%


 [OK] Successfully prefixed 1270 files.                                         

 // Memory usage: 21.40MB (peak: 64.68MB), time: 11.17s                         

composer --working-dir=build/set020-infection dump-autoload
Generating autoload files
Generated autoload files
# We generate the expected output file: we test that the scoping process
# does not alter it
php fixtures/set020-infection/vendor/infection/infection/bin/infection \
	--coverage=dist/infection-coverage \
	--skip-initial-tests \
	--only-covered \
	--no-progress
 1270/1270 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%PHP Fatal error:  Uncaught OutOfBoundsException: Package "infection/infection" is not installed in /home/runner/work/php-scoper/php-scoper/vendor/composer/InstalledVersions.php:928
Stack trace:
#0 /home/runner/work/php-scoper/php-scoper/vendor/composer/package-versions-deprecated/src/PackageVersions/Versions.php(155): Composer\InstalledVersions::getPrettyVersion()
#1 /home/runner/work/php-scoper/php-scoper/fixtures/set020-infection/vendor/infection/infection/src/Console/Application.php(75): PackageVersions\Versions::getVersion()
#2 /home/runner/work/php-scoper/php-scoper/fixtures/set020-infection/vendor/infection/infection/bin/infection(83): Infection\Console\Application->__construct()
#3 {main}
  thrown in /home/runner/work/php-scoper/php-scoper/vendor/composer/InstalledVersions.php on line 928
make: *** [e2e_020] Error 255
Makefile:204: recipe for target 'e2e_020' failed
</pre></details>

**Comment by theofidry** [comment]
Hm given how easy to is to upgrade to Composer 1 I wonder if it doesn't make sense to drop it completely...

**Comment by villfa** [comment]
Look here:
```
#0 /home/runner/work/php-scoper/php-scoper/vendor/composer/package-versions-deprecated/src/PackageVersions/Versions.php(155): Composer\InstalledVersions::getPrettyVersion()
```

This is not the right path. Instead it should be (with `build/set020-infection/`) :
```
/home/runner/work/php-scoper/php-scoper/build/set020-infection/vendor/composer/package-versions-deprecated/src/PackageVersions/Versions.php
```

I don't know more for now.

**Comment by smoench** [comment]
Thanks. Infection is running again but I'm not sure if it is correct.

**Comment by smoench** [comment]
I just parallelized the e2e tests in GitHub Actions 👀  (because composer v1 is not compatible with PHP 8.0)

**Comment by villfa** [comment]
> (because composer v1 is not compatible with PHP 8.0)

FYI, it will be compatible from the next release: https://github.com/composer/composer/issues/9523

File: .github/workflows/e2e.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                run: |
+                    source .composer-root-version
+                    echo "COMPOSER_ROOT_VERSION=${COMPOSER_ROOT_VERSION}" >> $GITHUB_ENV
+
+            -   name: Install dependencies
+                run: composer install --no-interaction --no-progress --no-suggest --prefer-dist
+
+            -   name: Run check-composer-root-version
+                run: make check-composer-root-version
+
```
Comment (villfa) [inline .github/workflows/e2e.yml:66]:
```suggestion
            -   name: install tree
                if: matrix.e2e == "e2e_032"
                run: sudo apt-get install -y tree

```

Comment (smoench) [reply .github/workflows/e2e.yml:69]: Thanks!

File: .github/workflows/generate-e2e-matrix.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
+<?php
```
Comment (villfa) [inline .github/workflows/generate-e2e-matrix.php:1]: if we go this way I would move this file to the level above since it's not a workflow file

Comment (smoench) [reply .github/workflows/generate-e2e-matrix.php:1]: I think we can just add the array in the workflow file to the matrix part and adding an include for the composer v1 e2e test.

Comment (smoench) [reply .github/workflows/generate-e2e-matrix.php]: Done

File: .github/workflows/e2e.yml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            -   name: Generate coverage
+                if: matrix.e2e == 'e2e_020'
+                run: make tc
```
Comment (villfa) [inline .github/workflows/e2e.yml:83]: We should not rely on the order of the tests to make them pass.
I propose you another solution: https://github.com/smoench/php-scoper/pull/1/commits/c270aa95321d4d1842c0c3f71e7108e1228a0cf0

Comment (smoench) [reply .github/workflows/e2e.yml:83]: Nice! I didn't notice you created a PR against my fork

Comment (smoench) [reply .github/workflows/e2e.yml:83]: I cherry-picked them. Thank you!

Comment (villfa) [reply .github/workflows/e2e.yml:83]: I just commited another fix for e2e_030

Comment (smoench) [reply .github/workflows/e2e.yml:83]: Are you planning to fix more e2e tests? I just want to prevent work is done twice :)

Comment (smoench) [reply .github/workflows/e2e.yml:83]: I added a task list -> https://github.com/humbug/php-scoper/pull/446#issue-529411440

Comment (villfa) [reply .github/workflows/e2e.yml:83]: Not right now. Maybe this evening if I have time.

**Comment by villfa** [comment]
I just found this: https://github.com/composer/composer/issues/9340#issuecomment-716210369
I think it's helpful to fix e2e_027.

**Comment by smoench** [comment]
I found a solution for e2e_027 (No, I didn't delete the fixture set):D
Laravel 5.6 is not compatible with PHP 8 so I decided run it only on PHP 7.3 with composer v1.

**Comment by smoench** [comment]
Next step would be reworking the build workflow and integrate e2e tests

**Comment by villfa** [comment]
I think we can do better for e2e_0210 and e2e_027.
I'm gonna try something and keep you posted.

File: Makefile [hunk 4]
```diff
@@ -230,6 +232,7 @@ e2e_020: $(PHPSCOPER) fixtures/set020-infection/vendor clover.xml
 .PHONY: e2e_0210
 e2e_0210: ## Run end-to-end tests for the fixture set 021 — Composer 1
 e2e_0210: $(PHPSCOPER) fixtures/set021-composer-1/vendor
+	rm -rf fixtures/set021-composer-1/vendor/*/*/tests
```
Comment (villfa) [inline Makefile:235]: @smoench now that Composer v1 is compatible with PHP8 we can remove this line

Comment (smoench) [reply Makefile]: Removed

**Comment by smoench** [comment]
@villfa Can you have a look at e2e_027 again? Still failing. Thanks in advance

**Comment by villfa** [comment]
> Can you have a look at e2e_027 again? Still failing. Thanks in advance

Yes, I will, but I can't tell you when exactly. Probably this weekend.

**Comment by villfa** [comment]
The remaining error is not directly related to PHP 8.

Here some scoped code (generated with PHP 7.3):
```php

if (!\function_exists('_PhpScoperd4c93190f897\\str_before')) {
    /**
     * Get the portion of a string before a given value.
     *
     * @param  string  $subject
     * @param  string  $search
     * @return string
     */
    function str_before($subject, $search)
    {
        return \_PhpScoperd4c93190f897\Illuminate\Support\Str::before($subject, $search);
    }
}
if (!\function_exists('str_contains')) {
    /**
     * Determine if a given string contains a given substring.
     *
     * @param  string  $haystack
     * @param  string|array  $needles
     * @return bool
     */
    function str_contains($haystack, $needles)
    {
        return \_PhpScoperd4c93190f897\Illuminate\Support\Str::contains($haystack, $needles);
    }
}
```

We see that `str_contains` has been treated as an internal function while this function doesn't exist before PHP 8.

A possible solution would be to take into account the `@since` annotation provided by phpstorm-stubs.

I'm not sure we want to fix this issue with this PR.

**Comment by smoench** [comment]
Well, this is more or less related to https://github.com/humbug/php-scoper/issues/440

**Comment by villfa** [comment]
@smoench e2e_027 is fixed with PHP8, and disabled with PHP7: https://github.com/smoench/php-scoper/pull/1

**Comment by theofidry** [comment]
Regarding the Symfony & Laravel test suite: if those are the blockers in the CI I would ignore them. There is definitely issues with those two but they are more related to polyfill support and whether or not a symbol is considered as internal depending of the target PHP versions, which are two sets of problems to be addressed on their own

**Comment by smoench** [comment]
Ready for review 🚀 

I also reworked the build workflow. It parallelized the e2e tests and is hopefully more readable.

**Comment by smoench** [comment]
A big thank you to @villfa. It was an honor to work with you on this!

**Review by theofidry — APPROVED** [review]
LGTM. Really impressed by the workflow it's really nice!

What is the issue about the pending build? Is it due to them being required by the branch in settings but no longer available?

**Comment by smoench** [comment]
Yes, the required builds are removed due to the rework and the [settings](https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/enabling-required-status-checks) needs to be updated.

**Comment by theofidry** [comment]
👌

**Comment by theofidry** [comment]
Thank you again @smoench @villfa for the amazing work

**Comment by TomasVotruba** [comment]
Hi, thank you for adding PHP 8!

It one of last blockers for Symplify packages, as we use scoper to release scoped PHP 8 packages.
Just curious, any ETA on release?

Thanks

**Comment by smoench** [comment]
@theofidry Can we have also a new release for php-scoper with PHP 8 support? Is there anything I can do to help you?

**Comment by theofidry** [comment]
https://github.com/humbug/php-scoper/releases/tag/0.14.0 released, sorry forgot about it :)

**Comment by smoench** [comment]
Thank you @theofidry  🚀

**Comment by TomasVotruba** [comment]
Thank you both!

**Pull request merged**